File: Features/notarychange-iou/workflows/src/main/java/net/corda/examples/notarychange/flows/SwitchNotaryFlow.java
package net.corda.examples.notarychange.flows;

import co.paralleluniverse.fibers.Suspendable;
import net.corda.core.contracts.StateAndRef;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.flows.*;
import net.corda.core.identity.Party;
import net.corda.core.node.services.Vault;
import net.corda.core.node.services.vault.QueryCriteria;
import net.corda.core.utilities.ProgressTracker;
import net.corda.examples.notarychange.states.IOUState;

import java.util.Collections;

@InitiatingFlow
@StartableByRPC
public class SwitchNotaryFlow extends FlowLogic<String> {

    private UniqueIdentifier linearId;
    private Party newNotary;

    public SwitchNotaryFlow(UniqueIdentifier linearId, Party newNotary) {
        this.linearId = linearId;
        this.newNotary = newNotary;
    }

    private final ProgressTracker.Step QUERYING_VAULT = new ProgressTracker.Step("Fetching IOU from node's vault.");
    private final ProgressTracker.Step INITITATING_TRANSACTION = new ProgressTracker.Step("Initiating Notary Change Transaction"){
        @Override
        public ProgressTracker childProgressTracker() {
            return AbstractStateReplacementFlow.Instigator.Companion.tracker();
        }
    };

    private final ProgressTracker progressTracker = new ProgressTracker(
            QUERYING_VAULT,
            INITITATING_TRANSACTION
    );

    @Override
    public ProgressTracker getProgressTracker() {
        return progressTracker;
    }

    @Override
    @Suspendable
    public String call() throws FlowException {
        progressTracker.setCurrentStep(QUERYING_VAULT);
        QueryCriteria queryCriteria = new QueryCriteria.LinearStateQueryCriteria(null, Collections.singletonList(linearId.getId()));
        Vault.Page results  = getServiceHub().getVaultService().queryBy(IOUState.class, queryCriteria);

        if(results.getStates().size() == 0){
            throw new FlowException("No IOU found for LinearId:" + linearId);
        }

        StateAndRef<IOUState> iouStateStateAndRef = (StateAndRef<IOUState>) results.getStates().get(0);
        progressTracker.setCurrentStep(INITITATING_TRANSACTION);
        subFlow(new NotaryChangeFlow<IOUState>(iouStateStateAndRef, newNotary, AbstractStateReplacementFlow.Instigator.Companion.tracker()));
        return "Notary Switched Successfully";
    }
}

File: kotlin-samples/Features/notarychange-iou/workflows/src/main/kotlin/net/corda/samples/notarychange/flows/SwitchNotaryFlow.kt
package net.corda.samples.notarychange.flows

import co.paralleluniverse.fibers.Suspendable
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.flows.*
import net.corda.core.flows.AbstractStateReplacementFlow.Instigator.Companion.tracker
import net.corda.core.identity.Party
import net.corda.core.node.services.vault.QueryCriteria
import net.corda.core.node.services.vault.QueryCriteria.LinearStateQueryCriteria
import net.corda.core.utilities.ProgressTracker
import net.corda.samples.notarychange.states.IOUState

@InitiatingFlow
@StartableByRPC
class SwitchNotaryFlow(private val linearId: UniqueIdentifier, private val newNotary: Party) : FlowLogic<String>() {
    private val QUERYING_VAULT = ProgressTracker.Step("Fetching IOU from node's vault.")
    private val INITITATING_TRANSACTION: ProgressTracker.Step = object : ProgressTracker.Step("Initiating Notary Change Transaction") {
        override fun childProgressTracker(): ProgressTracker? {
            return tracker()
        }
    }
    override val progressTracker = ProgressTracker(
            QUERYING_VAULT,
            INITITATING_TRANSACTION
    )

    @Suspendable
    @Throws(FlowException::class)
    override fun call(): String {
        progressTracker.currentStep = QUERYING_VAULT
        val queryCriteria: QueryCriteria = LinearStateQueryCriteria(null, listOf(linearId.id))
        val (states) = serviceHub.vaultService.queryBy(IOUState::class.java, queryCriteria)
        if (states.size == 0) {
            throw FlowException("No IOU found for LinearId:$linearId")
        }
        progressTracker.currentStep = INITITATING_TRANSACTION
        subFlow(NotaryChangeFlow(states[0], newNotary, tracker()))
        return "Notary Switched Successfully"
    }
}